File: app/src/main/java/com/dpis/module/TemplateWorkspaceBinder.java
package com.dpis.module;

import android.content.Context;
import android.content.SharedPreferences;
import android.view.View;
import android.widget.LinearLayout;
import android.widget.Toast;

import com.google.android.material.chip.ChipGroup;
import com.google.android.material.textview.MaterialTextView;

import java.util.List;
import java.util.ArrayList;
import java.util.Locale;

final class TemplateWorkspaceBinder {
    interface GlobalPrefillActions {
        void edit();
    }

    interface QuickTemplateActions {
        void apply(String templateId);

        void edit(String templateId);

        void select(String templateId);

        void create();

        void sort(List<QuickTemplateStore.QuickTemplate> templates);
    }

    private final Context context;
    private final SharedPreferences preferences;
    private final TemplateConfigSummaryFormatter formatter;
    private final TemplateSummaryChipBinder summaryChipBinder;
    private final QuickTemplateListAdapter quickTemplateListAdapter;
    private final GlobalPrefillActions globalPrefillActions;
    private final QuickTemplateActions quickTemplateActions;
    private static final float DISABLED_ACTION_ALPHA = 0.45f;

    TemplateWorkspaceBinder(Context context,
            GlobalPrefillActions globalPrefillActions,
            QuickTemplateActions quickTemplateActions) {
        this.context = context;
        this.globalPrefillActions = globalPrefillActions;
        this.quickTemplateActions = quickTemplateActions;
        this.preferences = context.getSharedPreferences(DpisConfigStore.GROUP, Context.MODE_PRIVATE);
        this.formatter = new TemplateConfigSummaryFormatter(
                new ResourceSummaryText(context),
                new TemplateTypefaceResolver(() -> ConfigStoreFactory.createLocalUiFontLibraryStore(
                        context, DpisApplication.getXposedService())));
        this.summaryChipBinder = new TemplateSummaryChipBinder(context);
        this.quickTemplateListAdapter = new QuickTemplateListAdapter(
                context,
                formatter,
                summaryChipBinder,
                this::onEditTemplate,
                this::onApplyTemplate,
                this::onSelectTemplate);
    }

    void bind(View workspaceView, String query) {
        if (workspaceView == null) {
            return;
        }
        bindGlobalPrefill(workspaceView);
        LinearLayout listContainer = workspaceView.findViewById(R.id.quick_template_list_container);
        MaterialTextView emptyState = workspaceView.findViewById(R.id.quick_template_empty_state);
        List<QuickTemplateStore.QuickTemplate> templates = new QuickTemplateStore(preferences).readAll();
        bindHeaderActions(workspaceView, templates);
        String normalizedQuery = normalizeQuery(query);
        boolean searching = !normalizedQuery.isEmpty();
        setVisible(workspaceView.findViewById(R.id.global_prefill_card), !searching);
        setVisible(workspaceView.findViewById(R.id.quick_template_section_header), !searching);
        quickTemplateListAdapter.bind(
                listContainer,
                emptyState,
                searching ? filterTemplates(templates, normalizedQuery) : templates,
                searching ? context.getString(R.string.quick_template_search_empty)
                        : context.getString(R.string.template_workspace_quick_templates_empty));
    }

    private void bindGlobalPrefill(View workspaceView) {
        TemplateConfigValue value = new GlobalPrefillStore(preferences).read();
        TemplateConfigSummaryFormatter.Result result = formatter.format(value);
        ChipGroup summaryChips = workspaceView.findViewById(R.id.global_prefill_summary_chips);
        MaterialTextView emptySummaryView = workspaceView.findViewById(R.id.global_prefill_empty_summary);
        View editButton = workspaceView.findViewById(R.id.global_prefill_edit_button);
        summaryChipBinder.bind(summaryChips, emptySummaryView, result);
        TouchFeedbackBinder.bindPressHaptic(editButton);
        editButton.setOnClickListener(v -> {
            if (globalPrefillActions != null) {
                globalPrefillActions.edit();
            }
        });
    }

    private void bindHeaderActions(View workspaceView, List<QuickTemplateStore.QuickTemplate> templates) {
        View sortButton = workspaceView.findViewById(R.id.quick_template_sort_button);
        if (sortButton != null) {
            boolean sortEnabled = templates != null && !templates.isEmpty();
            sortButton.setEnabled(sortEnabled);
            sortButton.setAlpha(sortEnabled ? 1f : DISABLED_ACTION_ALPHA);
            TouchFeedbackBinder.bindPressHaptic(sortButton);
            sortButton.setOnClickListener(v -> {
                if (quickTemplateActions != null) {
                    quickTemplateActions.sort(new QuickTemplateStore(preferences).readAll());
                }
            });
        }
        View createButton = workspaceView.findViewById(R.id.quick_template_create_button);
        if (createButton == null) {
            return;
        }
        TouchFeedbackBinder.bindPressHaptic(createButton);
        createButton.setOnClickListener(v -> {
            if (quickTemplateActions != null) {
                quickTemplateActions.create();
            }
        });
    }

    private void onEditTemplate(String templateId) {
        if (quickTemplateActions != null) {
            quickTemplateActions.edit(templateId);
        }
    }

    private void onApplyTemplate(String templateId) {
        if (quickTemplateActions != null) {
            quickTemplateActions.apply(templateId);
        }
    }

    private void onSelectTemplate(String templateId) {
        if (quickTemplateActions != null) {
            quickTemplateActions.select(templateId);
        }
    }

    private void showNotWiredToast() {
        Toast.makeText(context, R.string.template_workspace_not_wired, Toast.LENGTH_SHORT).show();
    }

    private static List<QuickTemplateStore.QuickTemplate> filterTemplates(
            List<QuickTemplateStore.QuickTemplate> templates,
            String normalizedQuery) {
        ArrayList<QuickTemplateStore.QuickTemplate> filtered = new ArrayList<>();
        if (templates == null) {
            return filtered;
        }
        for (QuickTemplateStore.QuickTemplate template : templates) {
            if (template.name.toLowerCase(Locale.ROOT).contains(normalizedQuery)) {
                filtered.add(template);
            }
        }
        return filtered;
    }

    private static String normalizeQuery(String query) {
        return query != null ? query.trim().toLowerCase(Locale.ROOT) : "";
    }

    private static void setVisible(View view, boolean visible) {
        if (view != null) {
            view.setVisibility(visible ? View.VISIBLE : View.GONE);
        }
    }

    private static final class ResourceSummaryText implements TemplateConfigSummaryFormatter.Text {
        private final Context context;

        ResourceSummaryText(Context context) {
            this.context = context;
        }

        @Override
        public String emptySummary() {
            return context.getString(R.string.template_workspace_summary_empty);
        }

        @Override
        public String viewportSummary(String detail) {
            return context.getString(R.string.template_workspace_summary_viewport, detail);
        }

        @Override
        public String viewportTargetTypeScale() {
            return context.getString(R.string.dialog_viewport_mode_system);
        }

        @Override
        public String viewportTargetTypeWidth() {
            return context.getString(R.string.dialog_viewport_mode_compat);
        }

        @Override
        public String fontSummary(String detail) {
            return context.getString(R.string.template_workspace_summary_font, detail);
        }

        @Override
        public String noValue() {
            return context.getString(R.string.app_status_no_value);
        }

        @Override
        public String typeface(String displayName) {
            return context.getString(R.string.template_workspace_summary_typeface, displayName);
        }

        @Override
        public String hookDomains() {
            return context.getString(R.string.template_workspace_summary_hook_domains);
        }

        @Override
        public String modeAuto() {
            return context.getString(R.string.template_workspace_mode_auto);
        }

        @Override
        public String modeSystem() {
            return context.getString(R.string.template_workspace_mode_system);
        }

        @Override
        public String modeCompat() {
            return context.getString(R.string.template_workspace_mode_compat);
        }
    }
}

File: app/src/main/java/com/dpis/module/templates/TemplateWorkspaceBinder.kt
package com.dpis.module.templates

import android.content.Context
import android.content.SharedPreferences
import android.view.View
import android.widget.LinearLayout
import com.dpis.module.ConfigStoreFactory
import com.dpis.module.DpisApplication
import com.dpis.module.DpisConfigStore
import com.dpis.module.R
import com.dpis.module.fonts.FontLibraryEntry
import com.dpis.module.fonts.FontLibraryStore
import com.dpis.module.ui.TouchFeedbackBinder
import com.google.android.material.chip.ChipGroup
import com.google.android.material.textview.MaterialTextView
import java.util.Locale

/** Binds the legacy template workspace and forwards user intent to the Activity host. */
class TemplateWorkspaceBinder(
    private val context: Context,
    private val globalPrefillActions: GlobalPrefillActions?,
    private val quickTemplateActions: QuickTemplateActions?
) {
    interface GlobalPrefillActions {
        fun edit()
    }

    interface QuickTemplateActions {
        fun apply(templateId: String)
        fun edit(templateId: String)
        fun select(templateId: String)
        fun create()
        fun sort(templates: List<QuickTemplateStore.QuickTemplate>)
    }

    private val preferences: SharedPreferences =
        context.getSharedPreferences(DpisConfigStore.GROUP, Context.MODE_PRIVATE)
    private val formatter = TemplateConfigSummaryFormatter(
        ResourceSummaryText(context),
        TemplateTypefaceResolver(::resolveImportedTypeface)
    )
    private val summaryChipBinder = TemplateSummaryChipBinder(context)
    private val quickTemplateListAdapter = QuickTemplateListAdapter(
        context,
        formatter,
        summaryChipBinder,
        ::onEditTemplate,
        ::onApplyTemplate,
        ::onSelectTemplate
    )

    fun bind(workspaceView: View?, query: String?) {
        if (workspaceView == null) return
        bindGlobalPrefill(workspaceView)
        val listContainer: LinearLayout = workspaceView.findViewById(R.id.quick_template_list_container)
        val emptyState: MaterialTextView = workspaceView.findViewById(R.id.quick_template_empty_state)
        val templates = QuickTemplateStore(context).readAll()
        bindHeaderActions(workspaceView, templates)
        val normalizedQuery = normalizeQuery(query)
        val searching = normalizedQuery.isNotEmpty()
        setVisible(workspaceView.findViewById(R.id.global_prefill_card), !searching)
        setVisible(workspaceView.findViewById(R.id.quick_template_section_header), !searching)
        quickTemplateListAdapter.bind(
            listContainer,
            emptyState,
            if (searching) filterTemplates(templates, normalizedQuery) else templates,
            if (searching) context.getString(R.string.quick_template_search_empty)
            else context.getString(R.string.template_workspace_quick_templates_empty)
        )
    }

    private fun bindGlobalPrefill(workspaceView: View) {
        val result = formatter.format(GlobalPrefillStore(preferences).read())
        val summaryChips: ChipGroup = workspaceView.findViewById(R.id.global_prefill_summary_chips)
        val emptySummaryView: MaterialTextView = workspaceView.findViewById(R.id.global_prefill_empty_summary)
        val editButton = workspaceView.findViewById<View>(R.id.global_prefill_edit_button)
        summaryChipBinder.bind(summaryChips, emptySummaryView, result)
        TouchFeedbackBinder.bindPressHaptic(editButton)
        editButton.setOnClickListener { globalPrefillActions?.edit() }
    }

    private fun resolveImportedTypeface(typefaceId: String): TemplateConfigSummaryFormatter.TypefaceStatus {
        val store: FontLibraryStore = ConfigStoreFactory.createLocalUiFontLibraryStore(
            context, DpisApplication.getXposedService()
        )
        val imported: FontLibraryEntry? = store.findById(typefaceId)
        return if (imported != null && store.resolveFontFile(typefaceId) != null) {
            TemplateConfigSummaryFormatter.TypefaceStatus.resolved(typefaceId, imported.displayName)
        } else {
            TemplateConfigSummaryFormatter.TypefaceStatus.missing(typefaceId)
        }
    }

    private fun bindHeaderActions(workspaceView: View, templates: List<QuickTemplateStore.QuickTemplate>) {
        workspaceView.findViewById<View>(R.id.quick_template_sort_button)?.let { sortButton ->
            val enabled = templates.isNotEmpty()
            sortButton.isEnabled = enabled
            sortButton.alpha = if (enabled) 1f else DISABLED_ACTION_ALPHA
            TouchFeedbackBinder.bindPressHaptic(sortButton)
            sortButton.setOnClickListener {
                quickTemplateActions?.sort(QuickTemplateStore(context).readAll())
            }
        }
        workspaceView.findViewById<View>(R.id.quick_template_create_button)?.let { createButton ->
            TouchFeedbackBinder.bindPressHaptic(createButton)
            createButton.setOnClickListener { quickTemplateActions?.create() }
        }
    }

    private fun onEditTemplate(templateId: String) = quickTemplateActions?.edit(templateId)
    private fun onApplyTemplate(templateId: String) = quickTemplateActions?.apply(templateId)
    private fun onSelectTemplate(templateId: String) = quickTemplateActions?.select(templateId)

    private fun filterTemplates(
        templates: List<QuickTemplateStore.QuickTemplate>,
        normalizedQuery: String
    ): List<QuickTemplateStore.QuickTemplate> = templates.filter {
        it.name.lowercase(Locale.ROOT).contains(normalizedQuery)
    }

    private fun normalizeQuery(query: String?) = query?.trim()?.lowercase(Locale.ROOT).orEmpty()

    private fun setVisible(view: View?, visible: Boolean) {
        view?.visibility = if (visible) View.VISIBLE else View.GONE
    }

    private class ResourceSummaryText(private val context: Context) : TemplateConfigSummaryFormatter.Text {
        override fun emptySummary() = context.getString(R.string.template_workspace_summary_empty)
        override fun viewportSummary(detail: String) = context.getString(R.string.template_workspace_summary_viewport, detail)
        override fun viewportTargetTypeScale() = context.getString(R.string.dialog_viewport_mode_system)
        override fun viewportTargetTypeWidth() = context.getString(R.string.dialog_viewport_mode_compat)
        override fun fontSummary(detail: String) = context.getString(R.string.template_workspace_summary_font, detail)
        override fun noValue() = context.getString(R.string.app_status_no_value)
        override fun typeface(displayName: String) = context.getString(R.string.template_workspace_summary_typeface, displayName)
        override fun hookDomains() = context.getString(R.string.template_workspace_summary_hook_domains)
        override fun modeAuto() = context.getString(R.string.template_workspace_mode_auto)
        override fun modeSystem() = context.getString(R.string.template_workspace_mode_system)
        override fun modeCompat() = context.getString(R.string.template_workspace_mode_compat)
    }

    private companion object {
        const val DISABLED_ACTION_ALPHA = 0.45f
    }
}
